transport/socks5/packet_dialer.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package socks5
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+type packetConn struct {
+	dstAddr net.Addr
+	pc      net.Conn
+	sc      transport.StreamConn
+}
+
+var _ net.Conn = (*packetConn)(nil)
+
+func (c *packetConn) LocalAddr() net.Addr {
+	// TODO: Is this right?
+	return c.pc.LocalAddr()
+}
+
+func (c *packetConn) RemoteAddr() net.Addr {
+	return c.dstAddr
+}
+
+func (c *packetConn) SetDeadline(t time.Time) error {
+	return c.pc.SetDeadline(t)
+}
+
+func (c *packetConn) SetReadDeadline(t time.Time) error {
+	return c.pc.SetReadDeadline(t)
+}
+
+func (c *packetConn) SetWriteDeadline(t time.Time) error {
+	return c.pc.SetWriteDeadline(t)
+}
+
+func (c *packetConn) Read(b []byte) (int, error) {
+	// TODO: read header
+	buffer := make([]byte, 65536) // Maximum size for UDP packet
+	n, err := c.pc.Read(buffer)
```

Comment:
Let's not read the entire packet. Instead, read each field as you go, like we do in the stream code.

Reply:
@fortuna Thanks for the feedback. I looked further into the suggested approach but it appears that I can only do one single Read and subsequent read operations time out. 

In UDP (User Datagram Protocol), each call to Read or ReadFrom on a UDP socket reads a single datagram from the network stack. This is fundamentally how UDP operates at the socket level; each datagram (packet) sent by a sender corresponds to one atomic message that should be read in its entirety by the receiver. If I attempt to read part of the message, I lose the rest of the message. This behaviour is different from TCP connection where I can read incrementally from the Connection. Since the package size is not known beforehand, I have to allocate maximum buffer size to fix the entire datagram.

Reply:
@fortuna some relevant discussion: https://forum.golangbridge.org/t/reading-from-udp-socket-blocks-even-if-there-is-data/3290/2

Reply:
I meant, we can wrap a buffer with Reader with `bytes.NewReader(buffer)`

Reply:
@fortuna got it. thanks for clarification. I implemented readAddress accordingly.